platform/linux-generic/odp_crypto.c

```diff
@@ -137,15 +139,56 @@ struct odp_crypto_global_s {
 	odp_spinlock_t                lock;
 	odp_crypto_generic_session_t *free;
 	odp_crypto_generic_session_t  sessions[MAX_SESSIONS];
+
+	/* These bitfields are cleared at alloc_session()
+	 * together with the rest of data */
+	bitset_t hmac_valid[ODP_THREAD_COUNT_MAX]
+		[(MAX_SESSIONS + ATOM_BITSET_SIZE - 1) / ATOM_BITSET_SIZE];
+	bitset_t cipher_valid[ODP_THREAD_COUNT_MAX]
+		[(MAX_SESSIONS + ATOM_BITSET_SIZE - 1) / ATOM_BITSET_SIZE];
+	bitset_t mac_cipher_valid[ODP_THREAD_COUNT_MAX]
+		[(MAX_SESSIONS + ATOM_BITSET_SIZE - 1) / ATOM_BITSET_SIZE];
+
 	odp_ticketlock_t              openssl_lock[0];
 };
 
 static odp_crypto_global_t *global;
 
+typedef struct crypto_local_t {
+	HMAC_CTX *hmac_ctx[MAX_SESSIONS];
+	EVP_CIPHER_CTX *cipher_ctx[MAX_SESSIONS];
+	EVP_CIPHER_CTX *mac_cipher_ctx[MAX_SESSIONS];
```

Comment:
This does not seem to be needed. There can be only cipher and mac in a session.

Reply:
Consider a case of encryption + AES-GMAC.

Reply:
ok, now I can see it. Ignore the previous comment.